File: store/src/test/java/com/nytimes/android/external/store3/StreamTest.java
package com.nytimes.android.external.store3;

import com.nytimes.android.external.store3.base.Fetcher;
import com.nytimes.android.external.store3.base.Persister;
import com.nytimes.android.external.store3.base.impl.BarCode;
import com.nytimes.android.external.store3.base.impl.Store;
import com.nytimes.android.external.store3.base.impl.StoreBuilder;

import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.mockito.runners.MockitoJUnitRunner;

import io.reactivex.Maybe;
import io.reactivex.Single;
import io.reactivex.observers.TestObserver;

import static org.mockito.Mockito.when;

@RunWith(MockitoJUnitRunner.class)
public class StreamTest {

    private static final String TEST_ITEM = "test";

    @Mock
    Fetcher<String, BarCode> fetcher;
    @Mock
    Persister<String, BarCode> persister;

    private final BarCode barCode = new BarCode("key", "value");

    private Store<String, BarCode> store;

    @Before
    public void setUp() {
        MockitoAnnotations.initMocks(this);

        store = StoreBuilder.<String>barcode()
                .persister(persister)
                .fetcher(fetcher)
                .open();

        when(fetcher.fetch(barCode))
                .thenReturn(Single.just(TEST_ITEM));

        when(persister.read(barCode))
                .thenReturn(Maybe.<String>empty())
                .thenReturn(Maybe.just(TEST_ITEM));

        when(persister.write(barCode, TEST_ITEM))
                .thenReturn(Single.just(true));
    }

    @Test
    public void testStream() {
        TestObserver<String> streamObservable = store.stream().test();
        streamObservable.assertValueCount(0);
        store.get(barCode).subscribe();
        streamObservable.assertValueCount(1);
    }

    @Test
    public void testStreamEmitsOnlyFreshData() {
        store.get(barCode).subscribe();
        TestObserver<String> streamObservable = store.stream().test();
        streamObservable.assertValueCount(0);
    }
}

File: store/src/test/java/com/nytimes/android/external/store3/StreamTest.kt
package com.nytimes.android.external.store3

import com.nhaarman.mockitokotlin2.mock
import com.nhaarman.mockitokotlin2.whenever
import com.nytimes.android.external.store3.base.Fetcher
import com.nytimes.android.external.store3.base.Persister
import com.nytimes.android.external.store3.base.impl.BarCode
import com.nytimes.android.external.store3.base.impl.StoreBuilder
import kotlinx.coroutines.runBlocking
import org.junit.Before
import org.junit.Test

class StreamTest {

    val fetcher: Fetcher<String, BarCode> = mock()
    val persister: Persister<String, BarCode> = mock()

    private val barCode = BarCode("key", "value")

    private val store = StoreBuilder.barcode<String>()
            .persister(persister)
            .fetcher(fetcher)
            .open()

    @Before
    fun setUp() = runBlocking<Unit> {
        whenever(fetcher.fetch(barCode)).thenReturn(TEST_ITEM)

        whenever(persister.read(barCode))
                .thenReturn(null)
                .thenReturn(TEST_ITEM)

        whenever(persister.write(barCode, TEST_ITEM))
                .thenReturn(true)
    }

    @Test
    fun testStream() = runBlocking<Unit> {
        val streamObservable = store.stream().test()
        streamObservable.assertValueCount(0)
        store.get(barCode)
        streamObservable.assertValueCount(1)
    }

    @Test
    fun testStreamEmitsOnlyFreshData() = runBlocking<Unit> {
        store.get(barCode)
        val streamObservable = store.stream().test()
        streamObservable.assertValueCount(0)
    }

    companion object {

        private val TEST_ITEM = "test"
    }
}
